### PR TITLE
Add support for the Host Power State feature.

### PR DESCRIPTION
**Pull request by pvanhorn**
This change modifies NcpBase, NcpBuffer, NcpUart and NcpSpi.

The purpose of this change is to allow NCP devices to wake a host that has entered into one of several Spinel defined low power states.  Depending on the particular State, different behaviors are required. For SPINEL_HOST_POWER_STATE_DEEP_SLEEP, the host must be awoken using otPlatWakeHost and for UART implementations, traffic to the host shall be deferred until any frame is received from the host.  For other low power states it is only necessary to call otPlatWakeHost.  

Any communication from the host after receiving SPINEL_PROP_HOST_POWER_STATE, will cause the NCP to transition the host to the SPINEL_HOST_POWER_STATE_ONLINE state thereby considering it to be capable of receiving any frame.

The rules for waking the host shall not be employed until after the NCP sends a response to the host's SPINEL_PROP_HOST_POWER_STATE command.

Currently, there is no frame filtering employed when the host is in one of the low power states.  It is understood that as part of a follow on Pull-Request frame filtering will be implemented so that particular frames, deemed to be of low priority, will be dropped instead of forwarding them to a sleeping host.

### CODE REVIEW

File: src/ncp/ncp_buffer.hpp [hunk 1]
```diff
@@ -234,6 +243,19 @@ class NcpFrameBuffer
      */
     uint16_t OutFrameGetLength(void);
 
+    /**
+     * This method provides a callback to NcpBuffer that it will use when it the current last frame
```
Comment (pvanhorn) [inline src/ncp/ncp_buffer.hpp]: done

File: src/ncp/ncp_buffer.hpp [hunk 2]
```diff
@@ -52,6 +52,15 @@ class NcpFrameBuffer
     typedef void (*BufferCallback)(void *aContext, NcpFrameBuffer *aNcpFrameBuffer);
 
     /**
+     * Defines a function pointer callback which is invoked to inform
```
Comment (pvanhorn) [inline src/ncp/ncp_buffer.hpp]: done

File: src/ncp/ncp_base.cpp [hunk 3]
```diff
@@ -3626,6 +3669,66 @@ ThreadError NcpBase::SetPropertyHandler_POWER_STATE(uint8_t header, spinel_prop_
...
+        }
+    }
+    else
+    {
+        errorCode = kThreadError_Parse;
+    }
+
+    if (errorCode == kThreadError_None)
+    {
+        errorCode = HandleCommandPropertyGet(header, key);
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: We probably need to handle the case if this returns `NoBuf` error (like `end of scan`)

**Comment by pvanhorn** [comment]
Satisfies issue #1771

File: src/ncp/ncp_base.cpp (hunk 3 above)
Comment (pvanhorn) [inline src/ncp/ncp_base.cpp]: Done.

**Comment by pvanhorn** [comment]
The build appears to be failing for reasons unrelated to the PR?

**Comment by jwhui** [comment]
@pvanhorn, the Travis failures are not specific to this PR.  We're actively looking into it.

File: src/ncp/ncp_base.cpp (hunk 3 above)
Comment (abtink) [inline src/ncp/ncp_base.cpp]: LGTM

**Comment by pvanhorn** [comment]
Tick tock.

**Pull request merged**